test/cpp/jit/test_gpu_validator.h

```diff
@@ -0,0 +1,321 @@
+          entry++;
+        }
+        double abs_tol = 0.0;
+        if (entry + 1 < sum_tolerance_entry.size()) {
+          // Grab the next entry up so we have some margin
+          abs_tol = sum_tolerance_entry[entry + 1][1];
+        } else {
+          // If we hit the end of the list, return twice the max error we
+          // measured
+          abs_tol = sum_tolerance_entry[sum_tolerance_entry.size() - 1][1] * 2.;
+        }
+        // Relative tol we're going to set to 1% of abs tol just for
+        // a small margin of rel error.
+        return {abs_tol, abs_tol * 0.01};
+      }
+    }
+    case DataType::Half: {
+      // Copied from float case
+      const auto& sum_tolerance_entry = tolerances.sum_tolerances_half;
+      const auto& base_abs = tolerances.base_half_abs_tol;
+      const auto& base_rel = tolerances.base_half_rel_tol;
+
+      if (reduction_size <= 1) {
+        // No reduction case
+        if (base_abs == -1 || base_rel == -1) {
+          return {sum_tolerance_entry[0][1], sum_tolerance_entry[1][1]};
+        } else {
+          return {base_abs, base_rel};
+        }
+      } else {
+        // Reduction case
+        size_t entry = 0;
+        while (sum_tolerance_entry[entry][0] < reduction_size &&
+               entry < sum_tolerance_entry.size()) {
+          entry++;
+        }
+        double abs_tol = 0.0;
+        if (entry + 1 < sum_tolerance_entry.size()) {
+          // Grab the next entry up so we have some margin
+          abs_tol = sum_tolerance_entry[entry + 1][1];
+        } else {
+          // If we hit the end of the list, return twice the max error we
+          // measured
+          abs_tol = sum_tolerance_entry[sum_tolerance_entry.size() - 1][1] * 2.;
+        }
+        // Relative tol we're going to set to 1% of abs tol just for
+        // a small margin of rel error.
+        return {abs_tol, abs_tol * 0.01};
+      }
+    }
+    case DataType::Int:
+      return {0.0, 0.0};
+    case DataType::Bool:
+      return {0.0, 0.0};
+    default:
+      TORCH_INTERNAL_ASSERT(
+          false, "Do not have tolerance computation for type ", dtype, ".");
+  }
+}
+
+class TORCH_CUDA_API ReductionSizeMapper : private IterVisitor {
+ public:
+  //! Runs through the fusion and determines how many reductions were performed
+  //! to compute each tensorview.
+  static std::unordered_map<TensorView*, int64_t> computeReductionSizes(
+      Fusion* fusion,
+      ExpressionEvaluator& expr_eval) {
+    ReductionSizeMapper mapper(fusion, expr_eval);
+    return mapper.reduction_map;
+  }
+
+ private:
+  ReductionSizeMapper(Fusion* fusion, ExpressionEvaluator& expr_eval)
+      : expr_eval_(expr_eval) {
+    // Initialize input values
+    for (auto inp : fusion->inputs()) {
+      if (inp->isA<TensorView>()) {
+        auto tv = inp->as<TensorView>();
+        // Shouldn't have any reductions, but run it through analysis anyways.
+        reduction_map[tv] = getReductionSize(tv);
```

Comment:
Quick question: we currently don't seem to have input tensors with reduction axes. Do we have more cases where we need to call `getReductionSize` on fusion inputs other than when we start partitioning fusions?

Reply:
I was thinking so and wanted it to be compatible with that yes.